### PR TITLE
Refactor GitHub Action test-and-lint workflow

### PR DESCRIPTION
**Pull request by dmholtz**
The refactoring introduces a matrix workflow with 2 DOF: OS and Python version. We avoid setting environment variables to make the workflow run on both Windows and Ubuntu in the same manner.

David Holtz <[david.holtz@mbition.io](mailto:david.holtz@mbition.io)>, on behalf of [MBition GmbH](https://mbition.io/).
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -1,28 +1,57 @@
...
+      - name: Install odxtools dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r requirements.txt
+      
+      # we need additional dependencies for testing and the type checker
+      # but we avoided to specify as hard dependencies within 'requirements.txt'
+      - name: Install test dependencies
+        run:  |
+          pip install types-Deprecated coveralls mypy pytest pytest-cov
```
Comment (kayoub5) [inline .github/workflows/tests.yml:37]: put them in a test-requirements.txt file

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -1,28 +1,57 @@
...
+      - name: Create somersault.pdx
+        run: |
+          python -m examples.mksomersaultpdx examples/somersault.pdx
+
+      - name: Run unit tests and report line coverage
+        # use `python -m pytest ...` instead of `pytest ...` to avoid setting the
+        # PYTHONPATH environment variable. This works independently of the OS.
+
+        # For some reason, pytest-cov reports invalid line coverage. As a remedy,
+        # we utilize the `coverage` module instead of `pytest-cov`. Line coverage
```
Comment (kayoub5) [inline .github/workflows/tests.yml:48]: * then why is pytest-cov in the dependency list if not used?
* sure this issue happens in latest version? pytest-cov uses the coverage module internally

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -1,28 +1,57 @@
...
 jobs:
   unittest-and-lint:
-    runs-on: ubuntu-latest
+
+    runs-on: ${{matrix.os}}
+
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest]
+        python-version: ["3.8", "3.10"]
```
Comment (kayoub5) [inline .github/workflows/tests.yml:18]: python 3.11 has been released

File: .github/workflows/tests.yml (hunk 2 above)
Comment (dmholtz) [reply .github/workflows/tests.yml:54]: Unfortunately, yes. `pytest-cov` reports only 36 % coverage in contrast to 65 % reported by `coverage`. For some reason, `pytest-cov` marks every statement on module level (i.e. imports, method signatures, constants) as uncovered, which seems wrong from my point of view.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (dmholtz) [reply .github/workflows/tests.yml:18]: Fair, I would suggest testing Python 3.8 and the latest version is then sufficient.

Comment (andlaus) [reply .github/workflows/tests.yml]: I agree

**Comment by andlaus** [comment]
merging. thanks for the review @kayoub5.

**Pull request merged**